nes-network/network/src/sender/sender.rs

```diff
@@ -0,0 +1,128 @@
+/*
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+use super::control::*;
+use crate::channel::Communication;
+use crate::protocol::*;
+use crate::sender::channel::{ChannelControlMessage, ChannelControlQueue};
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::runtime::Runtime;
+use tokio::sync::oneshot;
+use tracing::{Instrument, debug, info_span};
+
+pub struct SenderChannel {
+    queue: ChannelControlQueue,
+}
+
+pub enum TrySendDataResult {
+    Ok,
+    Full(TupleBuffer),
+    Closed(TupleBuffer),
+}
+impl SenderChannel {
+    pub fn try_send_data(&self, buffer: TupleBuffer) -> TrySendDataResult {
+        let result = self.queue.try_send(ChannelControlMessage::Data(buffer));
+        match result {
+            Ok(()) => TrySendDataResult::Ok,
+            Err(async_channel::TrySendError::Full(ChannelControlMessage::Data(buffer))) => {
+                TrySendDataResult::Full(buffer)
+            }
+            Err(async_channel::TrySendError::Closed(ChannelControlMessage::Data(buffer))) => {
+                TrySendDataResult::Closed(buffer)
+            }
+            _ => unreachable!(),
+        }
+    }
+    pub fn flush(&self) -> Result<bool> {
+        let (tx, rx) = oneshot::channel();
+        self.queue
+            .send_blocking(ChannelControlMessage::Flush(tx))
+            .map_err(|_| "Network Service Closed")?;
+        rx.blocking_recv()
+            .map_err(|_| "Network Service Closed".into())
+    }
+
+    pub fn close(self) -> bool {
+        self.queue.close()
+    }
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+pub type Error = Box<dyn std::error::Error + Send + Sync>;
```

Comment:
We could use anyhow/thiserror in the future to streamline the error types and give them more structure.